Add CLI actor delete/update and improve tombstone handling

pfefferle opened the pull request:
Introduces a new WP-CLI command to delete or update actors, with appropriate checks and activity queuing. Refactors the Actors class to allow bypassing user capability checks in retrieval methods. Moves tombstone creation for deleted objects to the outbox handler, and adds a test for the Tombstone::bury method. Removes redundant tombstone calls from the scheduler.

## Proposed changes:
- Added a new `actor` command to WP-CLI for deleting and updating actors
- Moved tombstone creation from the scheduler to the outbox handler for centralized processing
- Added test coverage for the `Tombstone::bury` method

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [X] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Added a new WP-CLI command to manage Actors.

</details>

pfefferle committed:
Add CLI actor delete/update and improve tombstone handling

Introduces a new WP-CLI command to delete or update actors, with appropriate checks and activity queuing. Refactors the Actors class to allow bypassing user capability checks in retrieval methods. Moves tombstone creation for deleted objects to the outbox handler, and adds a test for the Tombstone::bury method. Removes redundant tombstone calls from the scheduler.

matticbot committed:
Add changelog

pfefferle committed:
Add warning message before deleting Actor from Fediverse

A colorized warning is now displayed to inform users that deleting an Actor will only remove them from the Fediverse, not the database, and that the action may be irreversible.

obenland commented on includes/collection/class-actors.php:
```diff
@@ -59,16 +59,17 @@ class Actors {
 	 * @return Actor|User|Blog|Application|\WP_Error Actor object or WP_Error if not found or not permitted.
 	 */
-	public static function get_by_id( $user_id ) {
+	public static function get_by_id( $user_id, $bypass_user_can = false ) {
```
Can we add an `add_filter( 'activitypub_user_can_activitypub', '__return_true' );` in the CLI call instead?

pfefferle replied:
totally forgot about that filter! nice!

obenland commented on includes/class-cli.php:
```diff
@@ -395,6 +395,69 @@ public function comment( $args ) {
+				\WP_CLI::line( \WP_CLI::colorize( '%YThe Actor will not deleted from the database, but only from the Fediverse. Be aware that this action might be irreversible.%n' ) );
+				\WP_CLI::confirm( 'Do you really want to delete the Actor with the ID: ' . $args[1] );
+
+				$activity = new Activity();
+				$activity->set_actor( $actor->get_id() );
+				$activity->set_object( $actor->get_id() );
+				$activity->set_type( 'Delete' );
+
+				add_to_outbox( $activity, null, $args[1] );
```
It feels odd to me that we don't have an API for this. What makes calling `Actor::schedule_user_delete()` not a good option?

pfefferle replied:
This is exactly the reason why I created: https://github.com/Automattic/wordpress-activitypub/issues/2119

It is simply getting more and more confusing when a function is for local actor or remote actor handling.

There is a `delete` function for example that handles remote actor deletes.

pfefferle replied:
That is why I started simple and created the issue instead.

obenland replied:
Sounds good. In this case it's a local user, is this not what `Actor::schedule_user_delete()` handles?

pfefferle replied:
true!

obenland commented on includes/functions.php:
```diff
@@ -1536,6 +1536,11 @@ function add_to_outbox( $data, $activity_type = null, $user_id = 0, $content_vis
+	// Set Tombstones for deleted objects.
+	if ( 'Delete' === $activity->get_type() ) {
+		Tombstone::bury( object_to_uri( $activity->get_object() ) );
+	}
```
I don't know if we have a good place for this, but it feels like type-specific actions should be handled in a callback to `post_activitypub_add_to_outbox`, what do you think?

pfefferle replied:
Thought about that too.

pfefferle replied:
@obenland where should we put that hook? Is that part of the handler domain?

obenland replied:
Yeah no idea. The handlers are all inbox related currently, although I guess it doesn't have to stay that way?

Actually, nvm, Move already has an outbox callback. Let's put it in the handler, yeah.

pfefferle committed:
Remove bypass_user_can parameter from Actors methods

Eliminated the $bypass_user_can parameter from several methods in the Actors class, simplifying method signatures and always enforcing the user_can_activitypub check. This change improves code clarity and ensures consistent permission checks.

pfefferle committed:
Refactor CLI actor delete and update to use Scheduler

Replaces direct activity creation and outbox addition in the 'delete' and 'update' CLI commands with calls to Actor::schedule_user_delete and Actor::schedule_profile_update. This centralizes scheduling logic and simplifies the CLI command implementation.

pfefferle committed:
Temporarily allow ActivityPub for user delete CLI command

Adds a filter to always allow ActivityPub actions when scheduling a user delete via the CLI, ensuring the operation proceeds regardless of user settings. The filter is removed immediately after the operation to avoid side effects.

pfefferle committed:
Move Tombstone logic to Delete handler

Relocates the Tombstone creation for deleted objects from add_to_outbox in functions.php to a new post_add_to_outbox method in the Delete handler class. This change improves separation of concerns by centralizing delete-related logic.

pfefferle committed:
Add priority and args to post_add_to_outbox action

Updates the 'post_activitypub_add_to_outbox' action hook to specify a priority of 10 and accept 2 arguments. This allows the hooked method to receive additional parameters when triggered.

pfefferle committed:
Rename parameter in post_add_to_outbox method

Renamed the $outbox_activity_id parameter to $outbox_id in the post_add_to_outbox method and its docblock for clarity and consistency.

pfefferle committed:
Merge branch 'trunk' into add/actor-delete-cli

pfefferle committed:
Merge branch 'trunk' into add/actor-delete-cli

obenland reviewed (approved):
<img width="140" height="78" alt="Image" src="https://github.com/user-attachments/assets/8cf61642-d28b-4b33-a692-acd5f82036aa" />
